Add WeakListHashSet

rniwa opened the pull request:
#### a3d783a2cba3636a1e7094dd14c06f9428a4c60c
<pre>
Add WeakListHashSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=251181">https://bugs.webkit.org/show_bug.cgi?id=251181</a>

Reviewed by Chris Dumez.

Added WeakPtr version of ListHashSet.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/WeakListHashSet.h: Added.
(WTF::WeakListHashSet): Added.
* Source/WTF/wtf/WeakPtr.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::computeSizeOfWeakListHashSet):
(WTF_WeakPtr.WeakListHashSetBasic):
(WTF_WeakPtr.WeakListHashMapConstObjects):
(WTF_WeakPtr.WeakListHashSetRemoveIterator):
(WTF_WeakPtr.WeakListHashSetExpansion):
(WTF_WeakPtr.WeakListHashSetIsEmptyIgnoringNullReferences):
(WTF_WeakPtr.WeakListHashSetRemoveNullReferences):
(TestWebKitAPI::collectItemsInWeakListHashSet):
(WTF_WeakPtr.WeakListHashSetIterators):
(WTF_WeakPtr.WeakListHashSetAppendOrMoveToLast):
(WTF_WeakPtr.WeakListHashSetPrependOrMoveToFirst):

Canonical link: <a href="https://commits.webkit.org/259500@main">https://commits.webkit.org/259500@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f979c47cb64a1473c43eedfc711acb8b52d96c63

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14018 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114204 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174396 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15152 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4943 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97261 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113230 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110698 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11707 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94716 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39232 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93578 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80897 "Passed tests") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94795 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7362 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27696 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92826 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5093 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7455 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4285 "Found 1 new test failure: media/video-src-blob-replay.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30250 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13512 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47248 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101520 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9246 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25324 "Passed tests") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3493 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/WeakListHashSet.h:
```diff
@@ -0,0 +1,306 @@
+            return !iterator.get();
+        });
+        if (result)
+            amortizedCleanupIfNeeded(count);
```
I think this should be `increaseOperationCountSinceLastCleanup(count);` for consistency with other weak containers now.

rniwa replied:
Good point. Will fix.

cdumez reviewed (approved):
LGTM with fix.

rniwa committed:
Add WeakListHashSet
https://bugs.webkit.org/show_bug.cgi?id=251181

Reviewed by Chris Dumez.

Added WeakPtr version of ListHashSet.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/WeakListHashSet.h: Added.
(WTF::WeakListHashSet): Added.
* Source/WTF/wtf/WeakPtr.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::computeSizeOfWeakListHashSet):
(WTF_WeakPtr.WeakListHashSetBasic):
(WTF_WeakPtr.WeakListHashMapConstObjects):
(WTF_WeakPtr.WeakListHashSetRemoveIterator):
(WTF_WeakPtr.WeakListHashSetExpansion):
(WTF_WeakPtr.WeakListHashSetIsEmptyIgnoringNullReferences):
(WTF_WeakPtr.WeakListHashSetRemoveNullReferences):
(TestWebKitAPI::collectItemsInWeakListHashSet):
(WTF_WeakPtr.WeakListHashSetIterators):
(WTF_WeakPtr.WeakListHashSetAppendOrMoveToLast):
(WTF_WeakPtr.WeakListHashSetPrependOrMoveToFirst):

Canonical link: https://commits.webkit.org/259500@main
